File: ProgramacionUR/Preparciales/Preparcial2/Famas.cpp
#include <iostream>
#include <string>

using namespace std;

int main() {
    string numero_jugador1;

    // Jugador 1 introduce un número de 2 dígitos
    cout << "Jugador 1, introduce un número de 2 dígitos: ";
    cin >> numero_jugador1;

    // Verificar que el número tiene 2 dígitos
    if (numero_jugador1.length() != 2 || !isdigit(numero_jugador1[0]) || !isdigit(numero_jugador1[1])) {
        cout << "Número inválido. El juego ha terminado." << endl;
    } else {
        int intentos_restantes = 5;

        // Primer intento
        string numero_jugador2;
        cout << "Jugador 2, tienes " << intentos_restantes << " oportunidades. Introduce un número de 2 dígitos: ";
        cin >> numero_jugador2;
        if (numero_jugador2.length() == 2 && isdigit(numero_jugador2[0]) && isdigit(numero_jugador2[1])) {
            if (numero_jugador2 == numero_jugador1) {
                cout << "¡Has ganado!" << endl;
                return 0;
            } else {
                int puntos = (numero_jugador2[0] != numero_jugador1[0] && (numero_jugador2[0] == numero_jugador1[1])) +
                             (numero_jugador2[1] != numero_jugador1[1] && (numero_jugador2[1] == numero_jugador1[0]));
                int fama = (numero_jugador2[0] == numero_jugador1[0]) +
                           (numero_jugador2[1] == numero_jugador1[1]);
                if (fama > 0) {
                    cout << "Tienes " << fama << " fama" << (fama > 1 ? "s" : "") << "." << endl;
                }
                if (puntos > 0) {
                    cout << "Tienes " << puntos << " punto" << (puntos > 1 ? "s" : "") << "." << endl;
                }
                intentos_restantes--;
            }
        } else {
            cout << "Número inválido. Pasa al siguiente intento." << endl;
            intentos_restantes--;
        }

        // Segundo intento
        if (intentos_restantes > 0) {
            cout << "Jugador 2, tienes " << intentos_restantes << " oportunidades. Introduce un número de 2 dígitos: ";
            cin >> numero_jugador2;
            if (numero_jugador2.length() == 2 && isdigit(numero_jugador2[0]) && isdigit(numero_jugador2[1])) {
                if (numero_jugador2 == numero_jugador1) {
                    cout << "¡Has ganado!" << endl;
                    return 0;
                } else {
                    int puntos = (numero_jugador2[0] != numero_jugador1[0] && (numero_jugador2[0] == numero_jugador1[1])) +
                                 (numero_jugador2[1] != numero_jugador1[1] && (numero_jugador2[1] == numero_jugador1[0]));
                    int fama = (numero_jugador2[0] == numero_jugador1[0]) +
                               (numero_jugador2[1] == numero_jugador1[1]);
                    if (fama > 0) {
                        cout << "Tienes " << fama << " fama" << (fama > 1 ? "s" : "") << "." << endl;
                    }
                    if (puntos > 0) {
                        cout << "Tienes " << puntos << " punto" << (puntos > 1 ? "s" : "") << "." << endl;
                    }
                    intentos_restantes--;
                }
            } else {
                cout << "Número inválido. Pasa al siguiente intento." << endl;
                intentos_restantes--;
            }
        }

        // Tercer intento
        if (intentos_restantes > 0) {
            cout << "Jugador 2, tienes " << intentos_restantes << " oportunidades. Introduce un número de 2 dígitos: ";
            cin >> numero_jugador2;
            if (numero_jugador2.length() == 2 && isdigit(numero_jugador2[0]) && isdigit(numero_jugador2[1])) {
                if (numero_jugador2 == numero_jugador1) {
                    cout << "¡Has ganado!" << endl;
                    return 0;
                } else {
                    int puntos = (numero_jugador2[0] != numero_jugador1[0] && (numero_jugador2[0] == numero_jugador1[1])) +
                                 (numero_jugador2[1] != numero_jugador1[1] && (numero_jugador2[1] == numero_jugador1[0]));
                    int fama = (numero_jugador2[0] == numero_jugador1[0]) +
                               (numero_jugador2[1] == numero_jugador1[1]);
                    if (fama > 0) {
                        cout << "Tienes " << fama << " fama" << (fama > 1 ? "s" : "") << "." << endl;
                    }
                    if (puntos > 0) {
                        cout << "Tienes " << puntos << " punto" << (puntos > 1 ? "s" : "") << "." << endl;
                    }
                    intentos_restantes--;
                }
            } else {
                cout << "Número inválido. Pasa al siguiente intento." << endl;
                intentos_restantes--;
            }
        }

        // Cuarto intento
        if (intentos_restantes > 0) {
            cout << "Jugador 2, tienes " << intentos_restantes << " oportunidades. Introduce un número de 2 dígitos: ";
            cin >> numero_jugador2;
            if (numero_jugador2.length() == 2 && isdigit(numero_jugador2[0]) && isdigit(numero_jugador2[1])) {
                if (numero_jugador2 == numero_jugador1) {
                    cout << "¡Has ganado!" << endl;
                    return 0;
                } else {
                    int puntos = (numero_jugador2[0] != numero_jugador1[0] && (numero_jugador2[0] == numero_jugador1[1])) +
                                 (numero_jugador2[1] != numero_jugador1[1] && (numero_jugador2[1] == numero_jugador1[0]));
                    int fama = (numero_jugador2[0] == numero_jugador1[0]) +
                               (numero_jugador2[1] == numero_jugador1[1]);
                    if (fama > 0) {
                        cout << "Tienes " << fama << " fama" << (fama > 1 ? "s" : "") << "." << endl;
                    }
                    if (puntos > 0) {
                        cout << "Tienes " << puntos << " punto" << (puntos > 1 ? "s" : "") << "." << endl;
                    }
                    intentos_restantes--;
                }
            } else {
                cout << "Número inválido. Pasa al siguiente intento." << endl;
                intentos_restantes--;
            }
        }

        // Quinto intento
        if (intentos_restantes > 0) {
            cout << "Jugador 2, tienes " << intentos_restantes << " oportunidades. Introduce un número de 2 dígitos: ";
            cin >> numero_jugador2;
            if (numero_jugador2.length() == 2 && isdigit(numero_jugador2[0]) && isdigit(numero_jugador2[1])) {
                if (numero_jugador2 == numero_jugador1) {
                    cout << "¡Has ganado!" << endl;
                } else {
                    int puntos = (numero_jugador2[0] != numero_jugador1[0] && (numero_jugador2[0] == numero_jugador1[1])) +
                                 (numero_jugador2[1] != numero_jugador1[1] && (numero_jugador2[1] == numero_jugador1[0]));
                    int fama = (numero_jugador2[0] == numero_jugador1[0]) +
                               (numero_jugador2[1] == numero_jugador1[1]);
                    if (fama > 0) {
                        cout << "Tienes " << fama << " fama" << (fama > 1 ? "s" : "") << "." << endl;
                    }
                    if (puntos > 0) {
                        cout << "Tienes " << puntos << " punto" << (puntos > 1 ? "s" : "") << "." << endl;
                    }
                }
            } else {
                cout << "Número inválido. Pasa al siguiente intento." << endl;
            }
        }

        // Final del juego
        if (intentos_restantes == 0) {
            cout << "Se acabaron los intentos. El juego ha terminado." << endl;
        }
    }

    return 0;
}

File: ProgramacionUR/Preparciales/Preparcial2/Famas.py
# Jugador 1 introduce un número de 2 dígitos
numero_jugador1 = input("Jugador 1, introduce un número de 2 dígitos: ")

# Verificar que el número tiene 2 dígitos
if len(numero_jugador1) != 2 or not numero_jugador1.isdigit():
    print("Número inválido. El juego ha terminado.")
else:
    # Jugador 2 tiene 5 oportunidades para adivinar
    intentos_restantes = 5

    # Primer intento
    numero_jugador2 = input(f"Jugador 2, tienes {intentos_restantes} oportunidades. Introduce un número de 2 dígitos: ")
    if len(numero_jugador2) == 2 and numero_jugador2.isdigit():
        if numero_jugador2 == numero_jugador1:
            print("¡Has ganado!")
        else:
            puntos = 0
            fama = 0
            if numero_jugador2[0] == numero_jugador1[0]:
                fama += 1
            elif numero_jugador2[0] == numero_jugador1[1]:
                puntos += 1

            if numero_jugador2[1] == numero_jugador1[1]:
                fama += 1
            elif numero_jugador2[1] == numero_jugador1[0]:
                puntos += 1

            if fama > 0:
                print(f"Tienes {fama} fama{'s' if fama > 1 else ''}.")
            if puntos > 0:
                print(f"Tienes {puntos} punto{'s' if puntos > 1 else ''}.")
            intentos_restantes -= 1
    else:
        print("Número inválido. Pasa al siguiente intento.")
        intentos_restantes -= 1

    # Segundo intento
    if intentos_restantes > 0:
        numero_jugador2 = input(f"Jugador 2, tienes {intentos_restantes} oportunidades. Introduce un número de 2 dígitos: ")
        if len(numero_jugador2) == 2 and numero_jugador2.isdigit():
            if numero_jugador2 == numero_jugador1:
                print("¡Has ganado!")
            else:
                puntos = 0
                fama = 0
                if numero_jugador2[0] == numero_jugador1[0]:
                    fama += 1
                elif numero_jugador2[0] == numero_jugador1[1]:
                    puntos += 1

                if numero_jugador2[1] == numero_jugador1[1]:
                    fama += 1
                elif numero_jugador2[1] == numero_jugador1[0]:
                    puntos += 1

                if fama > 0:
                    print(f"Tienes {fama} fama{'s' if fama > 1 else ''}.")
                if puntos > 0:
                    print(f"Tienes {puntos} punto{'s' if puntos > 1 else ''}.")
                intentos_restantes -= 1
        else:
            print("Número inválido. Pasa al siguiente intento.")
            intentos_restantes -= 1

    # Tercer intento
    if intentos_restantes > 0:
        numero_jugador2 = input(f"Jugador 2, tienes {intentos_restantes} oportunidades. Introduce un número de 2 dígitos: ")
        if len(numero_jugador2) == 2 and numero_jugador2.isdigit():
            if numero_jugador2 == numero_jugador1:
                print("¡Has ganado!")
            else:
                puntos = 0
                fama = 0
                if numero_jugador2[0] == numero_jugador1[0]:
                    fama += 1
                elif numero_jugador2[0] == numero_jugador1[1]:
                    puntos += 1

                if numero_jugador2[1] == numero_jugador1[1]:
                    fama += 1
                elif numero_jugador2[1] == numero_jugador1[0]:
                    puntos += 1

                if fama > 0:
                    print(f"Tienes {fama} fama{'s' if fama > 1 else ''}.")
                if puntos > 0:
                    print(f"Tienes {puntos} punto{'s' if puntos > 1 else ''}.")
                intentos_restantes -= 1
        else:
            print("Número inválido. Pasa al siguiente intento.")
            intentos_restantes -= 1

    # Cuarto intento
    if intentos_restantes > 0:
        numero_jugador2 = input(f"Jugador 2, tienes {intentos_restantes} oportunidades. Introduce un número de 2 dígitos: ")
        if len(numero_jugador2) == 2 and numero_jugador2.isdigit():
            if numero_jugador2 == numero_jugador1:
                print("¡Has ganado!")
            else:
                puntos = 0
                fama = 0
                if numero_jugador2[0] == numero_jugador1[0]:
                    fama += 1
                elif numero_jugador2[0] == numero_jugador1[1]:
                    puntos += 1

                if numero_jugador2[1] == numero_jugador1[1]:
                    fama += 1
                elif numero_jugador2[1] == numero_jugador1[0]:
                    puntos += 1

                if fama > 0:
                    print(f"Tienes {fama} fama{'s' if fama > 1 else ''}.")
                if puntos > 0:
                    print(f"Tienes {puntos} punto{'s' if puntos > 1 else ''}.")
                intentos_restantes -= 1
        else:
            print("Número inválido. Pasa al siguiente intento.")
            intentos_restantes -= 1

    # Quinto intento
    if intentos_restantes > 0:
        numero_jugador2 = input(f"Jugador 2, tienes {intentos_restantes} oportunidades. Introduce un número de 2 dígitos: ")
        if len(numero_jugador2) == 2 and numero_jugador2.isdigit():
            if numero_jugador2 == numero_jugador1:
                print("¡Has ganado!")
            else:
                puntos = 0
                fama = 0
                if numero_jugador2[0] == numero_jugador1[0]:
                    fama += 1
                elif numero_jugador2[0] == numero_jugador1[1]:
                    puntos += 1

                if numero_jugador2[1] == numero_jugador1[1]:
                    fama += 1
                elif numero_jugador2[1] == numero_jugador1[0]:
                    puntos += 1

                if fama > 0:
                    print(f"Tienes {fama} fama{'s' if fama > 1 else ''}.")
                if puntos > 0:
                    print(f"Tienes {puntos} punto{'s' if puntos > 1 else ''}.")
        else:
            print("Número inválido. Pasa al siguiente intento.")
            intentos_restantes -= 1

    # Final del juego
    if intentos_restantes == 0:
        print("Se acabaron los intentos. El juego ha terminado.")